File: sdk/inferlet/javascript/src/grammar.ts
// Grammar compilation + incremental matching — `pie:inferlet/grammar`.
//
// `Grammar` compiles a JSON Schema / regex / EBNF source once for the bound
// model's vocabulary; `Matcher` walks one generation against it, exposing
// the packed allowed-token bitmask that `mask.ts` interprets.

import * as _grammar from 'pie:inferlet/grammar@0.3.0';

/** A grammar the host refused to compile, or a token the matcher refused. */
export class GrammarError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'GrammarError';
  }
}

function wit<T>(what: string, f: () => T): T {
  try {
    return f();
  } catch (e: unknown) {
    const payload = (e as { payload?: unknown })?.payload;
    throw new GrammarError(`${what}: ${typeof payload === 'string' ? payload : String(e)}`);
  }
}

/** A compiled constraint grammar. */
export class Grammar {
  /** @internal */
  constructor(readonly inner: _grammar.Grammar) {}

  /** A grammar for JSON values conforming to `schema`. */
  static fromJsonSchema(schema: string): Grammar {
    return new Grammar(wit('grammar from JSON schema', () => _grammar.Grammar.fromJsonSchema(schema)));
  }

  /** The grammar of any JSON value. */
  static json(): Grammar {
    return new Grammar(_grammar.Grammar.json());
  }

  static fromRegex(pattern: string): Grammar {
    return new Grammar(wit('grammar from regex', () => _grammar.Grammar.fromRegex(pattern)));
  }

  static fromEbnf(ebnf: string): Grammar {
    return new Grammar(wit('grammar from EBNF', () => _grammar.Grammar.fromEbnf(ebnf)));
  }

  toString(): string {
    return this.inner.toString();
  }
}

/** An incremental match of one generation against a `Grammar`. */
export class Matcher {
  private readonly inner: _grammar.Matcher;

  constructor(grammar: Grammar | _grammar.Matcher) {
    this.inner = grammar instanceof Grammar ? new _grammar.Matcher(grammar.inner) : grammar;
  }

  /** Advance the match by `tokenIds`; a forbidden token throws `GrammarError`. */
  acceptTokens(tokenIds: ArrayLike<number>): void {
    wit('accept tokens', () => this.inner.acceptTokens(Uint32Array.from(tokenIds)));
  }

  /** The packed allowed-token bitmask for the next position (see `mask.ts`). */
  mask(): Uint32Array {
    return this.inner.mask();
  }

  isTerminated(): boolean {
    return this.inner.isTerminated();
  }

  reset(): void {
    this.inner.reset();
  }

  /** An independent copy at the current position (branching). */
  fork(): Matcher {
    return new Matcher(this.inner.fork());
  }

  rollback(numTokens: number): void {
    this.inner.rollback(numTokens);
  }

  rollbackCapacity(): number {
    return this.inner.rollbackCapacity();
  }
}

File: sdk/inferlet/python/src/inferlet/grammar.py
"""
Grammar compilation + incremental matching — ``pie:inferlet/grammar``.

:class:`Grammar` compiles a JSON Schema / regex / EBNF source once for the
bound model's vocabulary; :class:`Matcher` walks one generation against it,
exposing the packed allowed-token bitmask that :mod:`inferlet.mask`
interprets::

    from inferlet import grammar, mask
    from inferlet.eta import *

    constraint = grammar.Matcher(grammar.Grammar.from_json_schema(schema))
    allowed = Channel([vocab], dtype.bool)
    allowed.put(mask.unpack_mask(constraint.mask(), vocab))
    ...
    constraint.accept_tokens([token])
"""

from __future__ import annotations

from typing import Sequence

from componentize_py_types import Err as _WitErr
from wit_world.imports import grammar as _grammar


class GrammarError(Exception):
    """A grammar the host refused to compile, or a token the matcher refused."""


def _wit(what: str, fn, *args):
    try:
        return fn(*args)
    except _WitErr as e:
        raise GrammarError(f"{what}: {e.value}") from None


class Grammar:
    """A compiled constraint grammar."""

    __slots__ = ("_inner",)

    def __init__(self, inner) -> None:
        self._inner = inner

    @classmethod
    def from_json_schema(cls, schema: str) -> "Grammar":
        """A grammar for JSON values conforming to ``schema``."""
        return cls(_wit("grammar from JSON schema", _grammar.Grammar.from_json_schema, schema))

    @classmethod
    def json(cls) -> "Grammar":
        """The grammar of any JSON value."""
        return cls(_grammar.Grammar.json())

    @classmethod
    def from_regex(cls, pattern: str) -> "Grammar":
        return cls(_wit("grammar from regex", _grammar.Grammar.from_regex, pattern))

    @classmethod
    def from_ebnf(cls, ebnf: str) -> "Grammar":
        return cls(_wit("grammar from EBNF", _grammar.Grammar.from_ebnf, ebnf))

    def __str__(self) -> str:
        return self._inner.to_string()


class Matcher:
    """An incremental match of one generation against a :class:`Grammar`."""

    __slots__ = ("_inner",)

    def __init__(self, grammar: Grammar) -> None:
        self._inner = _grammar.Matcher(grammar._inner)

    @classmethod
    def _wrap(cls, inner) -> "Matcher":
        m = cls.__new__(cls)
        m._inner = inner
        return m

    def accept_tokens(self, token_ids: Sequence[int]) -> None:
        """Advance the match by ``token_ids``; a token the grammar forbids
        raises :class:`GrammarError`."""
        _wit("accept tokens", self._inner.accept_tokens, [int(t) for t in token_ids])

    def mask(self) -> list[int]:
        """The packed allowed-token bitmask for the next position (see
        :mod:`inferlet.mask`)."""
        return list(self._inner.mask())

    def is_terminated(self) -> bool:
        return self._inner.is_terminated()

    def reset(self) -> None:
        self._inner.reset()

    def fork(self) -> "Matcher":
        """An independent copy at the current position (branching)."""
        return Matcher._wrap(self._inner.fork())

    def rollback(self, num_tokens: int) -> None:
        self._inner.rollback(num_tokens)

    def rollback_capacity(self) -> int:
        return self._inner.rollback_capacity()
